[14.0][ADD] hr_timesheet_predefined_description: Add predefined descriptions for timesheets

juanjosesegui-tecnativa opened the pull request:
@Tecnativa TT50619

juanjosesegui-tecnativa committed:
[ADD] hr_timesheet_predefined_description: Add predefined descriptions for timesheets

TT50619

Co-Authored-By: Pedro M. Baeza <pedro.baeza@tecnativa.com>

carlos-lopez-tecnativa commented on hr_timesheet_predefined_description/README.rst:
```diff
@@ -0,0 +1,130 @@
+=============
+
+To use this module, you need to:
+
+#. Navigate to *Timesheets* > *Configuration* > *Predefined Descriptions*.
+   * Click on *Create* to add a new predefined description.
+   * Fill in the description name that you want to make available for timesheet entries.
+   * Save your changes.
+
+#. **Associate Predefined Descriptions with Timesheets:**
+
+   * When filling out timesheets, users can select a predefined description from the dropdown list in the *Predefined Description* field.
+   * The selected description will automatically populate the *Description* field in the timesheet entry.
+
+#. **Modify Predefined Descriptions:**
+
+   * If you need to update or delete a predefined description, you can do so from the same *Predefined Descriptions* menu.
+   * Ensure that changes to predefined descriptions are reflected correctly in any associated timesheet entries.
+
+Usage
+=====
+
+Once the `hr_timesheet_predefined_description` module is installed, you can start using predefined descriptions in your timesheets as follows:
+
+#. **Create Predefined Descriptions:**
+
+   - Navigate to *Timesheets > Configuration > Predefined Descriptions*.
+   - Click on *Create* to add a new predefined description.
+   - Enter the name of the description and save.
+
+#. **Use Predefined Descriptions in Timesheets:**
+
+   - When creating or editing a timesheet entry, you will see a new field labeled *Predefined Description*.
+   - Select a predefined description from the dropdown menu.
+   - The *Description* field will be automatically populated with the selected predefined description.
+
+#. **Modify Timesheet Descriptions:**
+
+   - If needed, you can manually override the description after selecting a predefined one.
+   - The predefined descriptions can be edited or removed from the *Predefined Descriptions* menu.
```
I see that `Configuration` and `Usage` are very similar, especially regarding the creation of new predefined descriptions. Is it possible to be more specific in each section, or should we remove it from Usage if it is explained in Configuration?

pedrobaeza replied:
That should go only on configuration.

carlos-lopez-tecnativa commented on hr_timesheet_predefined_description/tests/test_account_analytic_line.py:
```diff
@@ -0,0 +1,37 @@
+        self.PredefinedDescription = self.env["timesheet.predefined.description"]
+        self.AccountAnalyticLine = self.env["account.analytic.line"]
+
+        # Crear una descripción predefinida para usar en las pruebas
```
Please write in English

carlos-lopez-tecnativa commented on hr_timesheet_predefined_description/tests/test_account_analytic_line.py:
```diff
@@ -0,0 +1,37 @@
+            {"name": "Test Predefined Description"}
+        )
+
+        # Crear una línea analítica
```
and here

carlos-lopez-tecnativa commented on hr_timesheet_predefined_description/tests/test_account_analytic_line.py:
```diff
@@ -0,0 +1,37 @@
+        )
+
+    def test_onchange_predefined_description(self):
+        self.analytic_line.predefined_description_id = self.predefined_description
```
To test `onchange`, I think it is better to use the `Form` class.

carlos-lopez-tecnativa commented on hr_timesheet_predefined_description/tests/test_hr_timesheet_predefined_description.py:
```diff
@@ -0,0 +1,20 @@
+
+    def test_create_predefined_description_with_empty_name(self):
+        with self.assertRaises(ValidationError):
+            self.PredefinedDescription.create({"name": ""})
```
I don't see a use case in this test. What is the purpose?

carlos-lopez-tecnativa commented on hr_timesheet_predefined_description/views/hr_timesheet_views.xml:
```diff
@@ -0,0 +1,15 @@
+                <field name="predefined_description_id" />
+            </field>
+        </field>
+    </record>
```
Is it possible to add the new field to the `Timesheet` page in the Project Task?

![image](https://github.com/user-attachments/assets/6aa42e2d-9840-4679-a793-81791e7fdd02)

carlos-lopez-tecnativa commented on hr_timesheet_predefined_description/models/hr_timesheet_predefined_description.py:
```diff
@@ -0,0 +1,18 @@
+    name = fields.Char(string="Description", required=True)
+
+    @api.model
+    def create(self, vals):
+        if not vals.get("name"):
+            raise ValidationError(_("The field 'Description' cannot be empty."))
+        return super(TimesheetPredefinedDescription, self).create(vals)
```
If the field is `required`, this code is unnecessary because Odoo can't save a record if any required field is empty.

pedrobaeza replied:
That should be in custom, not in this module, and it should check if it has the default value ("/") as well.

carlos-lopez-tecnativa commented on hr_timesheet_predefined_description/models/hr_timesheet_predefined_description.py:
```diff
@@ -0,0 +1,18 @@
+from odoo.exceptions import ValidationError
+
+
+class TimesheetPredefinedDescription(models.Model):
+    _name = "timesheet.predefined.description"
+    _description = "Predefined Description for Timesheets"
+
+    name = fields.Char(string="Description", required=True)
+
```
I think it’s a good option to add the `company_id` field(no required) and the respective record rules. What do you think, or is it out of scope for the module?

pedrobaeza replied:
Yes, please.

victoralmau commented on hr_timesheet_predefined_description/security/ir.model.access.csv:
```diff
@@ -0,0 +1,2 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_timesheet_predefined_description,timesheet.predefined.description,model_timesheet_predefined_description,hr_timesheet.group_hr_timesheet_user,1,1,1,1
```
```suggestion
access_timesheet_predefined_description_user,timesheet.predefined.description.user,model_timesheet_predefined_description,hr_timesheet.group_hr_timesheet_user,1,0,0,0
access_timesheet_predefined_description_admin,timesheet.predefined.description.admin,model_timesheet_predefined_description,hr_timesheet.group_timesheet_manager,1,1,1,1
```

victoralmau commented on hr_timesheet_predefined_description/views/hr_timesheet_views.xml:
```diff
@@ -0,0 +1,37 @@
+                <attribute name="invisible">1</attribute>
+            </field>
+            <field name="name" position="after">
+                <field name="predefined_description_id" />
```
```suggestion
                <field name="predefined_description_id" required="True" />
```
If we hide the name, we must make the `predefined_description_id` field required (only in view). Same comment in task view.

pedrobaeza replied:
As said before, this shouldn't be enforced here in this module, but on custom part. The OCA part must include the m2o, but let the free text one as it was.

juanjosesegui-tecnativa replied:
Hello, I have removed the invisible from the name field and added optional="hide" so that both fields can be used freely. Is this correct? @pedrobaeza

victoralmau reviewed (changes requested):
Can you add as demo data several predefined descriptions?

pedrobaeza commented on hr_timesheet_predefined_description/models/account_analytic_line.py:
```diff
@@ -0,0 +1,25 @@
+    )
+
+    def write(self, vals):
+        if "predefined_description_id" in vals:
```
Only do that if no name is specified:

```suggestion
        if "predefined_description_id" in vals and (vals.get("name") or "/") == "/":
```

pedrobaeza commented on hr_timesheet_predefined_description/models/account_analytic_line.py:
```diff
@@ -0,0 +1,25 @@
+        comodel_name="timesheet.predefined.description", string="Predefined Description"
+    )
+
+    def write(self, vals):
```
Always document the overriden method for expressing the intention of the code.

```suggestion
    def write(self, vals):
        # Fill the predefined description if no description provided
```

pedrobaeza commented on hr_timesheet_predefined_description/data/hr_timesheet_predefined_description_data.xml:
This must be demo data and follow the model name, so rename it to `timesheet_predefined_description_demo.xml`

pedrobaeza commented on hr_timesheet_predefined_description/readme/CONFIGURE.rst:
```diff
@@ -0,0 +1,15 @@
+To configure the `hr_timesheet_predefined_description` module, follow these steps:
+
+#. **Add Predefined Descriptions:**
+
+   - Go to *Timesheets* > *Configuration* > *Predefined Descriptions*.
+   - Click *Create*, enter the description name, and save.
+
+#. **Modify Predefined Descriptions:**
+
+   - Update or delete existing descriptions from the same menu as needed.
+
+#. **Ensure Changes Reflect in Timesheets:**
+
+   - Any changes to predefined descriptions should be verified to reflect in timesheet entries.
+
```
Take care of the proper RST syntax:

```suggestion
**Add Predefined Descriptions:**

#. Go to *Timesheets* > *Configuration* > *Predefined Descriptions*.
#. Click on *Create*.
#. Enter the description name.
#. In a multi-company environment, check the company visibility.
#. Save.

**Modify Predefined Descriptions:**

#. Go to *Timesheets* > *Configuration* > *Predefined Descriptions*.
#. Click on an existing description.
#. Type the new description.

NOTE: These changes are not retro-actively applied to existing timesheets.
```

pedrobaeza commented on hr_timesheet_predefined_description/readme/USAGE.rst:
```diff
@@ -0,0 +1,11 @@
+To use predefined descriptions in your timesheets:
```
This file should contain the usage of the module, not the configuration. That's it how you select the predefined description in the timesheets, and they are reflected in the description.

pedrobaeza commented on hr_timesheet_predefined_description/tests/__init__.py:
```diff
@@ -0,0 +1,4 @@
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from . import test_hr_timesheet_predefined_description
+from . import test_test_account_analytic_line
```
Incorrect import (and sort them):

```suggestion
from . import test_account_analytic_line
```

pedrobaeza commented on hr_timesheet_predefined_description/tests/test_account_analytic_line.py:
```diff
@@ -0,0 +1,37 @@
+
+class TestAccountAnalyticLine(common.TransactionCase):
+    def setUp(self):
+        super(TestAccountAnalyticLine, self).setUp()
```
```suggestion
        super().setUp()
```

pedrobaeza commented on hr_timesheet_predefined_description/tests/test_account_analytic_line.py:
```diff
@@ -0,0 +1,37 @@
+from odoo.tests import Form, common
+
+
+class TestAccountAnalyticLine(common.TransactionCase):
```
On v14, you should declare it as `common.SavepointCase` and override `setUpClass`, not `setUp`.

pedrobaeza commented on hr_timesheet_predefined_description/tests/test_hr_timesheet_predefined_description.py:
```diff
@@ -0,0 +1,15 @@
+# Copyright 2024 Tecnativa - Juan José Seguí
```
Why 2 files? Everything can be done in one class, but several `def test_...` methods.

pedrobaeza commented on hr_timesheet_predefined_description/views/hr_timesheet_predefined_description_views.xml:
```diff
@@ -0,0 +1,26 @@
+        <field name="model">timesheet.predefined.description</field>
+        <field name="arch" type="xml">
+            <tree string="Predefined Descriptions" editable="bottom">
+                <field name="name" />
```
Missing `company_id` field, and also missing record rules for multi-company.

pedrobaeza commented on hr_timesheet_predefined_description/views/hr_timesheet_views.xml:
```diff
@@ -0,0 +1,40 @@
+        <field name="model">account.analytic.line</field>
+        <field name="inherit_id" ref="hr_timesheet.hr_timesheet_line_tree" />
+        <field name="arch" type="xml">
+            <field name="name" position="attributes">
```
Don't hide the name field. Make it readonly (with `force_save="1"`) if the predefined description is set.

pedrobaeza commented on hr_timesheet_predefined_description/views/hr_timesheet_views.xml:
```diff
@@ -0,0 +1,40 @@
+            <field name="name" position="attributes">
+                <attribute name="invisible">1</attribute>
+            </field>
+            <field name="name" position="after">
```
```suggestion
            <field name="name" position="before">
```

pedrobaeza commented on hr_timesheet_predefined_description/views/hr_timesheet_views.xml:
```diff
@@ -0,0 +1,40 @@
+            </field>
+        </field>
+    </record>
+    <record id="view_project_task_form_inherit" model="ir.ui.view">
```
The same as above for this view.

juanjosesegui-tecnativa commented:
I have implemented all the requested changes and adjusted the tests so that they pass without any errors. If you notice any further modifications are needed, feel free to let me know.

victoralmau reviewed (approved):
Code and functional review OK

pedrobaeza commented:
/ocabot merge nobump
